src/MongoDB.Driver.Core/Core/Bindings/CoreSession.cs

```diff
@@ -425,17 +427,45 @@ private void EnsureStartTransactionCanBeCalled()
         {
             if (_currentTransaction == null)
             {
-                return;
+                EnsureTransactionsAreSupported();
             }
+            else
+            {
+                switch (_currentTransaction.State)
+                {
+                    case CoreTransactionState.Aborted:
+                    case CoreTransactionState.Committed:
+                        break;
 
-            switch (_currentTransaction.State)
+                    default:
+                        throw new InvalidOperationException("Transaction already in progress.");
+                }
+            }
+        }
+
+        private void EnsureTransactionsAreSupported()
+        {
+            var connectedDataBearingServers = _cluster.Description.Servers.Where(s => s.State == ServerState.Connected && s.IsDataBearing).ToList();
+
+            if (connectedDataBearingServers.Count == 0)
             {
-                case CoreTransactionState.Aborted:
-                case CoreTransactionState.Committed:
-                    return;
+                throw new NotSupportedException("StartTransaction cannot determine if transactions are supported because there are no connected servers.");
+            }
 
-                default:
-                    throw new InvalidOperationException("Transaction already in progress.");
+            foreach (var connectedDataBearingServer in connectedDataBearingServers)
+            {
+                Feature.Transactions.ThrowIfNotSupported(connectedDataBearingServer.Version);
+                if (connectedDataBearingServer.Type == ServerType.ShardRouter)
+                {
+                    if (connectedDataBearingServer.Version < new SemanticVersion(4, 2, 0))
```

Comment:
can we use something like this:

>Feature.###.IsSupported(connectedDataBearingServer.Version)

?

Reply:
Normally yes... and in master we do.

But this is on the v2.8.x branch and we don't have Feature.ShardedTransactions on this branch and I didn't think we wanted to add it (on this branch).